solr/core/src/java/org/apache/solr/update/DirectUpdateHandler2.java

```diff
@@ -417,7 +417,8 @@ private void addAndDelete(AddUpdateCommand cmd, List<UpdateLog.DBQ> deletesAfter
   }
 
   private Term getIdTerm(AddUpdateCommand cmd) {
```

Comment:
I think it's trappy/dangerous to invoke getDocsList() which is an innocent looking getter that actually flattens the input and caches it rendering any possible further changes to the unflattend docs silently ignored.  Can you change this method to accept the List of flattened documents as its argument?  And change getDocsList to not cache the result and be named something like computeFlattenedDocs()

Reply:
Latest diff shows this logic calls cmd.computeFlattenedDocs() and we definitely don't want this method computing that!
Now that I look at the code in my IDE, I can appreciate that this is a bit of a tricky issue though, since neither callers of getIdTerm yet have the List<SolrInputDocument> to give to this method.  Hmmmm.   Maybe `updateDocument` should not take an updateTerm as an argument but should instead figure it out and return it?  Note that if isInPlaceUpdate, the updateTerm will always be the unique key field; you needn't check for a list of docs.

Reply:
After looking at the code, it seems like updateDocOrDocValues has to return the idTerm to be used by the caller, and the updateTerm can then be figured out by the caller, either the cmd.updateTerm or the idTerm.
I will push a commit soon, hopefully the code can do the talking.